# Naive Bayes

In [90]:
import csv

In [126]:
arquivo='data/tempo.csv'
arquivo='data/tenis.csv'
data = []
with open(arquivo) as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=';')
    count = 0
    for row in csv_reader:
        if count != 0:
            data.append(tuple(row))
        count += 1
print(len(data))
print(data)

14
[('sol', 'quente', 'alta', 'falso', 'não'), ('sol', 'quente', 'alta', 'verdadeiro', 'não'), ('nublado', 'quente', 'alta', 'falso', 'sim'), ('chuva', 'agradável', 'alta', 'falso', 'sim'), ('chuva', 'fria', 'normal', 'falso', 'sim'), ('chuva', 'fria', 'normal', 'verdadeiro', 'não'), ('nublado', 'fria', 'normal', 'verdadeiro', 'sim'), ('sol', 'agradável', 'alta', 'falso', 'não'), ('sol', 'fria', 'normal', 'falso', 'sim'), ('chuva', 'agradável', 'normal', 'falso', 'sim'), ('sol', 'agradável', 'normal', 'verdadeiro', 'sim'), ('nublado', 'agradável', 'alta', 'verdadeiro', 'sim'), ('nublado', 'quente', 'normal', 'falso', 'sim'), ('chuva', 'agradável', 'alta', 'verdadeiro', 'não')]


In [137]:
data

[('sol', 'quente', 'alta', 'falso', 'não'),
 ('sol', 'quente', 'alta', 'verdadeiro', 'não'),
 ('nublado', 'quente', 'alta', 'falso', 'sim'),
 ('chuva', 'agradável', 'alta', 'falso', 'sim'),
 ('chuva', 'fria', 'normal', 'falso', 'sim'),
 ('chuva', 'fria', 'normal', 'verdadeiro', 'não'),
 ('nublado', 'fria', 'normal', 'verdadeiro', 'sim'),
 ('sol', 'agradável', 'alta', 'falso', 'não'),
 ('sol', 'fria', 'normal', 'falso', 'sim'),
 ('chuva', 'agradável', 'normal', 'falso', 'sim'),
 ('sol', 'agradável', 'normal', 'verdadeiro', 'sim'),
 ('nublado', 'agradável', 'alta', 'verdadeiro', 'sim'),
 ('nublado', 'quente', 'normal', 'falso', 'sim'),
 ('chuva', 'agradável', 'alta', 'verdadeiro', 'não')]

In [114]:
len(data)

14

In [115]:
total = len(data)
cdata = {}
pdata = {}
combdata = {}
pcombdata = {}
for d in data:
    if d in combdata:
        combdata[d] += 1
    else:
        combdata[d] = 1
    for i, key in enumerate(d):
        if key in cdata:
            cdata[key] += 1 
        else:
            cdata[key] = 1
        pdata[key] = cdata[key] / total

for comb in combdata:
    pcombdata[comb] = combdata[comb] / cdata[comb[1]] * pdata[comb[1]] / pdata[comb[0]]

In [ ]:
total = len(data)
cdata = {}
pdata = {}
combdata = {}
pcombdata = {}
for d in data:
    if d in combdata:
        combdata[d] += 1
    else:
        combdata[d] = 1
    for i, key in enumerate(d):
        if key in cdata:
            cdata[key] += 1 
        else:
            cdata[key] = 1
        pdata[key] = cdata[key] / total

for comb in combdata:
    pcombdata[comb] = combdata[comb] / cdata[comb[1]] * pdata[comb[1]] / pdata[comb[0]]

In [160]:
total = len(data)
classes = list(set([x[-1] for x in data]))
combdata = {}
pdata = {}
for row in data:
    classe = (row[-1])
    if classe in combdata:
        combdata[classe] += 1
    else:
        combdata[classe] = 1    
    for d in row[:-1]:
        if d in combdata:
            combdata[d] += 1
        else:
            combdata[d] = 1
        key = (d,classe)
        if key in combdata:
            combdata[key] += 1
        else:
            combdata[key] = 1
combdata


{'não': 5,
 'sol': 5,
 ('sol', 'não'): 3,
 'quente': 4,
 ('quente', 'não'): 2,
 'alta': 7,
 ('alta', 'não'): 4,
 'falso': 8,
 ('falso', 'não'): 2,
 'verdadeiro': 6,
 ('verdadeiro', 'não'): 3,
 'sim': 9,
 'nublado': 4,
 ('nublado', 'sim'): 4,
 ('quente', 'sim'): 2,
 ('alta', 'sim'): 3,
 ('falso', 'sim'): 6,
 'chuva': 5,
 ('chuva', 'sim'): 3,
 'agradável': 6,
 ('agradável', 'sim'): 4,
 'fria': 4,
 ('fria', 'sim'): 3,
 'normal': 7,
 ('normal', 'sim'): 6,
 ('chuva', 'não'): 2,
 ('fria', 'não'): 1,
 ('normal', 'não'): 1,
 ('verdadeiro', 'sim'): 3,
 ('agradável', 'não'): 2,
 ('sol', 'sim'): 2}

In [166]:
probdata = {}
for c in combdata:
    if type(c) == tuple:
        probdata[c] = combdata[c] / combdata[c[1]]
    else:
        probdata[c] = combdata[c] / total
probdata['classes'] = classes
probdata

{'não': 0.35714285714285715,
 'sol': 0.35714285714285715,
 ('sol', 'não'): 0.6,
 'quente': 0.2857142857142857,
 ('quente', 'não'): 0.4,
 'alta': 0.5,
 ('alta', 'não'): 0.8,
 'falso': 0.5714285714285714,
 ('falso', 'não'): 0.4,
 'verdadeiro': 0.42857142857142855,
 ('verdadeiro', 'não'): 0.6,
 'sim': 0.6428571428571429,
 'nublado': 0.2857142857142857,
 ('nublado', 'sim'): 0.4444444444444444,
 ('quente', 'sim'): 0.2222222222222222,
 ('alta', 'sim'): 0.3333333333333333,
 ('falso', 'sim'): 0.6666666666666666,
 'chuva': 0.35714285714285715,
 ('chuva', 'sim'): 0.3333333333333333,
 'agradável': 0.42857142857142855,
 ('agradável', 'sim'): 0.4444444444444444,
 'fria': 0.2857142857142857,
 ('fria', 'sim'): 0.3333333333333333,
 'normal': 0.5,
 ('normal', 'sim'): 0.6666666666666666,
 ('chuva', 'não'): 0.4,
 ('fria', 'não'): 0.2,
 ('normal', 'não'): 0.2,
 ('verdadeiro', 'sim'): 0.3333333333333333,
 ('agradável', 'não'): 0.4,
 ('sol', 'sim'): 0.2222222222222222,
 'classes': ['sim', 'não']}

In [191]:
def verificar(probdata, tupla):
    probClasse = {}
    valClasse = 0
    maxClasse = ''
    for classe in probdata['classes']:
        numerador = 1
        denominador = 1
        for v in tupla:
            key = (v, classe)
            numerador = numerador * probdata[key]
            denominador = denominador * probdata[v]
        numerador = numerador * probdata[classe]
        probClasse[classe] = numerador / denominador
        print(classe,probClasse[classe])
        if probClasse[classe] > valClasse:
            valClasse = probClasse[classe]
            maxClasse = classe
    return(maxClasse)


In [192]:
tupla = ('sol', 'fria', 'alta', 'verdadeiro')
print(verificar(probdata, tupla))

sim 0.2419753086419753
não 0.9408
não


In [193]:
tupla = ('chuva', 'fria', 'normal', 'verdadeiro')
print(verificar(probdata, tupla))

sim 0.7259259259259259
não 0.15680000000000005
sim


In [123]:
for comb in combdata:
    print(comb)
    for c in comb[:-1]:
        print(c)
        print(comb[-1])
    
    pcombdata[comb] = combdata[comb] / cdata[comb[1]] * pdata[comb[1]] / pdata[comb[0]]

('sunny', 'no')
sunny
no
('overcast', 'yes')
overcast
yes
('rainny', 'yes')
rainny
yes
('sunny', 'yes')
sunny
yes
('rainny', 'no')
rainny
no


In [120]:
pcombdata

{('sunny', 'no'): 0.4,
 ('overcast', 'yes'): 1.0,
 ('rainny', 'yes'): 0.39999999999999997,
 ('sunny', 'yes'): 0.6,
 ('rainny', 'no'): 0.6}

In [116]:
combdata

{('sunny', 'no'): 2,
 ('overcast', 'yes'): 4,
 ('rainny', 'yes'): 2,
 ('sunny', 'yes'): 3,
 ('rainny', 'no'): 3}

In [117]:
cdata

{'sunny': 5, 'no': 5, 'overcast': 4, 'yes': 9, 'rainny': 5}

In [118]:
pdata

{'sunny': 0.35714285714285715,
 'no': 0.35714285714285715,
 'overcast': 0.2857142857142857,
 'yes': 0.6428571428571429,
 'rainny': 0.35714285714285715}

In [119]:
pcombdata

{('sunny', 'no'): 0.4,
 ('overcast', 'yes'): 1.0,
 ('rainny', 'yes'): 0.39999999999999997,
 ('sunny', 'yes'): 0.6,
 ('rainny', 'no'): 0.6}

Problem: Players will play if weather is sunny. Is this statement is correct?

We can solve it using above discussed method of posterior probability.

P(Yes | Sunny) = P( Sunny | Yes) * P(Yes) / P (Sunny)

Here we have P (Sunny |Yes) = 3/9 = 0.33, P(Sunny) = 5/14 = 0.36, P( Yes)= 9/14 = 0.64

Now, P (Yes | Sunny) = 0.33 * 0.64 / 0.36 = 0.60, which has higher probability.